app/internal/page/data.go

```diff
@@ -161,42 +161,80 @@ func readDate(t time.Time) Date {
 	}
 }
 
+type WitnessCode struct {
+	Code    string
+	Created time.Time
+}
+
+func (w *WitnessCode) HasExpired() bool {
+	return w.Created.Before(time.Now().Add(-30 * time.Minute))
+}
+
 type LpaStore interface {
-	Get(context.Context, string) (*Lpa, error)
-	Put(context.Context, string, *Lpa) error
+	Create(context.Context) (*Lpa, error)
+	GetAll(context.Context) (*Lpa, error)
+	Get(context.Context) (*Lpa, error)
+	Put(context.Context, *Lpa) error
 }
 
 type lpaStore struct {
 	dataStore DataStore
 	randomInt func(int) int
 }
 
-type WitnessCode struct {
-	Code    string
-	Created time.Time
+type sessionData struct {
+	SessionID string
+	LpaID     string
 }
 
-func (w *WitnessCode) HasExpired() bool {
-	return w.Created.Before(time.Now().Add(-30 * time.Minute))
+func (d *sessionData) String() string {
+	if d == nil {
+		return "sessionData<nil>"
+	}
+
+	return fmt.Sprintf("sessionData(sessionID=%s lpaID=%s)", d.SessionID, d.LpaID)
+}
+
+func sessionDataFromContext(ctx context.Context) *sessionData {
+	data, _ := ctx.Value((*sessionData)(nil)).(*sessionData)
```

Comment:
Whats the purpose of the `(nil)` here?

Reply:
I'm creating a `nil` of type `*sessionData`. So I get a unique key for the context, using nil which means it doesn't allocate. It's a one line way of doing

```
var key *sessionData
data, _ := ctx.Value(key).(*sessionData)
```

Reply:
Right, making sense now. It hadn't clicked that `nil` was being used as the value. So would this also be valid (not that it would make sense for what we need here):

```go
data, _ := ctx.Value(&sessionData{someProp: 'someValue'}).(*sessionData)
```

Happy to keep it as is now I can see the purpose.

Reply:
that would be valid if you'd set something with the same key